<a href="https://colab.research.google.com/github/jrg94/CSE5522/blob/lab3/lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CSE 5522 - Lab 3
By Jeremy Grifski

In this lab, we'll take a look at Hidden Markov Models (HMMs) for the Eisner Ice Cream Problem. 

## Part 1: Viterbi Algorithm

Implement the Viterbi algorithm for HMMs for Eisner's Ice Cream Problem (predict whether each day is hot or cold based on the number of ice creams eaten).  Remember that the Viterbi algorithm computes the most likely sequence for an input.

Your solution should be able to handle variable length sequences (in the range of 3-5).

[This zip file has observation probabilities, transition probabilities, and test data for evaluation](https://osu.instructure.com/courses/76815/files/18485497/download).  Please read the probabilities and observations from a file, do not hard-code them. (This is so that we can test with different data/probabilities.)

The observation and transition probabilities have rows being the variable of interest, and columns being the conditioning variables.    For example, P(2|H) is in the 3rd row (including header), 3rd column (including row label).  The columns sum to 1.

The test data has one line per sequence.  When a sequence is less than five observations long, the last columns are filled with zeros.

**1.0**: Let's setup the environment for data loading.

In [0]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

**1.1**: Now, we'll need to load all the data from the CSV files.

In [0]:
observation_dataframe = pd.read_csv("observationProbs.csv")
test_dataframe = pd.read_csv("testData.csv")
transition_dataframe = pd.read_csv("transitionProbs.csv")

**1.2**: Let's now take a peak at our data.

In [1518]:
display(
    observation_dataframe.shape,
    observation_dataframe.head(), 
    test_dataframe.shape, 
    test_dataframe.head(),
    transition_dataframe.shape,
    transition_dataframe.head()
)

(3, 3)

P(x|...)       C       H
0         1  0.6407  0.0002
1         2  0.1481  0.5341
2         3  0.2122  0.4657

(10, 6)

SeqNumber  Obs1  Obs2  Obs3  Obs4  Obs5
0          1     2     3     3     2     3
1          2     2     3     2     2     0
2          3     3     1     3     3     1
3          4     2     1     1     0     0
4          5     1     1     1     2     3

(3, 4)

P(x|...)     C     H  START
0        C  0.86  0.07    0.5
1        H  0.07  0.86    0.5
2     STOP  0.07  0.07    0.0

**1.3**: Now, let's build up some strings for indexing the data sets.

In [0]:
COND_PROB_LABEL = "P(x|...)"
HOT_LABEL = "H"
COLD_LABEL = "C"
START_LABEL = "START"
SEQUENCE_NUM_LABEL = "SeqNumber"
LENGTH_LABEL = "length"

**1.4**: With our data loaded, we can begin to construct our M and C matrices—assuming the first sequence for now. Here, we will assume 0 is cold and 1 is hot. This assumption matches the input data tables.  

In [1520]:
test_dataframe_copy = test_dataframe.copy()

# Gets everything but seqnumber and adds column with sequence length
cols = test_dataframe_copy.loc[:, test_dataframe_copy.columns != SEQUENCE_NUM_LABEL]  
test_dataframe_copy[LENGTH_LABEL] = cols.astype(bool).sum(axis=1)  

# Computes dimensions of m matrix
m_height = test_dataframe_copy.loc[test_dataframe_copy[SEQUENCE_NUM_LABEL] == 1, LENGTH_LABEL].iloc[0]
m_width = observation_dataframe.shape[1] - 1

# Creates m and c matrix
m = np.zeros(shape=(m_height, m_width))
c = np.zeros(shape=(m_height - 1, m_width))

display(m, c)

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

**1.5**: At this point, we can initialize the first row of the m matrix using the following formula: M<sub>1, k</sub> = π<sub>k</sub>B<sub>k,y<sub>1</sub></sub>. Here, π represents the prior probabilities and B is the emission probability.

We can get π from the START column of our transition matrix. Meanwhile, we can get B from a sequence in our test data and our observation data.  

In [1521]:
# Compute prior probabilities from transition matrix
PRIOR_PROB_HOT = transition_dataframe.loc[transition_dataframe[COND_PROB_LABEL] == HOT_LABEL, START_LABEL].iloc[0]
PRIOR_PROB_COLD = transition_dataframe.loc[transition_dataframe[COND_PROB_LABEL] == COLD_LABEL, START_LABEL].iloc[0]

# Compute observation 1 from test data 
obs1 = test_dataframe_copy.loc[test_dataframe_copy[SEQUENCE_NUM_LABEL] == 1, cols.columns[0]].iloc[0]

# Compute emission probability from observation 1
P_OBS_GIVEN_HOT = observation_dataframe.loc[observation_dataframe[COND_PROB_LABEL] == obs1, HOT_LABEL].iloc[0]
P_OBS_GIVEN_COLD = observation_dataframe.loc[observation_dataframe[COND_PROB_LABEL] == obs1, COLD_LABEL].iloc[0]

# Compute initial probabilities
m11 = PRIOR_PROB_COLD * P_OBS_GIVEN_COLD
m12 = PRIOR_PROB_HOT * P_OBS_GIVEN_HOT

display(
    obs1,
    PRIOR_PROB_HOT,
    PRIOR_PROB_COLD,
    P_OBS_GIVEN_HOT,
    P_OBS_GIVEN_COLD,
    m11,
    m12
)

2

0.5

0.5

0.5341

0.1481

0.07405

0.26705

**1.6**: Let's find a way to turn this initialization step into a function.

In [0]:
def prior_probability(transition: pd.DataFrame, given_label: str):
  """
  A helper function which returns the prior probability of a label
  """
  return transition.loc[transition[COND_PROB_LABEL] == given_label, START_LABEL].iloc[0]

In [0]:
def emission_probability(observation: pd.DataFrame, given_label: str, obs: str):
  """
  A helper function which returns the conditional probablity of an observation
  given a label
  """
  return observation.loc[observation[COND_PROB_LABEL] == obs, given_label].iloc[0]

In [0]:
def test_observation(test: pd.DataFrame, cols: pd.Series, sequence: int, obs: int):
  """
  A helper function which returns an observation given a sequence number and
  an index.
  """
  return test.loc[test[SEQUENCE_NUM_LABEL] == sequence, cols.columns[obs]].iloc[0]

In [0]:
def init_matrices(test, observation, transition, sequence):
  test_copy = test.copy()

  # Gets everything but seqnumber and adds column with sequence length
  COLS = test_copy.loc[:, test_copy.columns != SEQUENCE_NUM_LABEL]  
  test_copy[LENGTH_LABEL] = COLS.astype(bool).sum(axis=1)  

  # Computes dimensions of m matrix
  M_HEIGHT = test_copy.loc[test_copy[SEQUENCE_NUM_LABEL] == sequence, LENGTH_LABEL].iloc[0]
  M_WIDTH = observation.shape[1] - 1

  # Creates m and c matrix
  m = np.zeros(shape=(M_HEIGHT, M_WIDTH))
  c = np.zeros(shape=(M_HEIGHT - 1, M_WIDTH))

  # Compute prior probabilities from transition matrix
  PRIOR_PROB_COLD = prior_probability(transition, COLD_LABEL)
  PRIOR_PROB_HOT = prior_probability(transition, HOT_LABEL)

  # Compute observation 1 from test data 
  OBS1 = test_observation(test_copy, COLS, sequence, 0)

  # Compute emission probability from observation 1
  P_OBS1_GIVEN_COLD = emission_probability(observation, COLD_LABEL, OBS1)
  P_OBS1_GIVEN_HOT = emission_probability(observation, HOT_LABEL, OBS1)

  # Compute initial probabilities
  M11 = PRIOR_PROB_COLD * P_OBS1_GIVEN_COLD
  M12 = PRIOR_PROB_HOT * P_OBS1_GIVEN_HOT

  m[0, 0] = M11
  m[0, 1] = M12

  return m, c

In [1526]:
m, c = init_matrices(test_dataframe, observation_dataframe, transition_dataframe, 1)
display(m, c)

array([[0.07405, 0.26705],
       [0.     , 0.     ],
       [0.     , 0.     ],
       [0.     , 0.     ],
       [0.     , 0.     ]])

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

**1.7**: With an initialization function, we can begin tackling the next step: writing a step function for each row of m and c.

In [0]:
def transition_probability(transition: pd.DataFrame, label: str, given_label: str):
  """
  A helper function which returns the probability of a label today given
  a label yesterday.
  """
  return transition.loc[transition[COND_PROB_LABEL] == label, given_label].iloc[0]

In [0]:
def compute_state(p_max_list: list, p_obs_given_state: float):
  """
  A helper function which computes the most probable state by finding the maximum
  probability from a list of probabilities.
  """
  INDEX = np.argmax(p_max_list)
  P_STATE_TODAY = p_obs_given_state * p_max_list[INDEX]
  return INDEX, P_STATE_TODAY

In [0]:
def step(m, c, test, observation, transition, sequence, obs_index):
  """
  A helper function which updates the m and c matrices for a given observation.
  """
  COLS = test.loc[:, test.columns != SEQUENCE_NUM_LABEL]  
  OBS = test_observation(test, COLS, sequence, obs_index)

  # Compute emissions probability: B
  P_OBS_GIVEN_COLD = emission_probability(observation, COLD_LABEL, OBS)
  P_OBS_GIVEN_HOT = emission_probability(observation, HOT_LABEL, OBS)

  # Compute transition probability: A
  P_C_TODAY_GIVEN_C_YESTERDAY = transition_probability(transition, COLD_LABEL, COLD_LABEL)
  P_C_TODAY_GIVEN_H_YESTERDAY = transition_probability(transition, COLD_LABEL, HOT_LABEL)
  P_H_TODAY_GIVEN_C_YESTERDAY = transition_probability(transition, HOT_LABEL, COLD_LABEL)
  P_H_TODAY_GIVEN_H_YESTERDAY = transition_probability(transition, HOT_LABEL, HOT_LABEL)

  # Compute probabilities from previous day
  P_COLD_YESTERDAY = m[obs_index - 1, 0]
  P_HOT_YESTERDAY = m[obs_index - 1, 1]

  # Setup the list of cold today related probabilities
  COLD_PROBS = [
      P_C_TODAY_GIVEN_C_YESTERDAY * P_COLD_YESTERDAY,
      P_C_TODAY_GIVEN_H_YESTERDAY * P_HOT_YESTERDAY
  ]

  HOT_PROBS = [
      P_H_TODAY_GIVEN_C_YESTERDAY * P_COLD_YESTERDAY,
      P_H_TODAY_GIVEN_H_YESTERDAY * P_HOT_YESTERDAY
  ]

  # Compute index and probability for c and m
  C_COLD_VAL, M_COLD_VAL = compute_state(COLD_PROBS, P_OBS_GIVEN_COLD)
  C_HOT_VAL, M_HOT_VAL = compute_state(HOT_PROBS, P_OBS_GIVEN_HOT)

  # Populate c and m
  c[obs_index - 1, :] = C_COLD_VAL, C_HOT_VAL
  m[obs_index, :] = M_COLD_VAL, M_HOT_VAL

In [1530]:
step(m, c, test_dataframe, observation_dataframe, transition_dataframe, 1, 1)
display(m, c)

array([[0.07405   , 0.26705   ],
       [0.01351353, 0.10695406],
       [0.        , 0.        ],
       [0.        , 0.        ],
       [0.        , 0.        ]])

array([[0., 1.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

**1.8**: With the step function written, we can continually generate rows for every observation. 

In [1531]:
for i in range(1, m.shape[0]):
  step(m, c, test_dataframe, observation_dataframe, transition_dataframe, 1, i)
display(m, c)

array([[0.07405   , 0.26705   ],
       [0.01351353, 0.10695406],
       [0.00246611, 0.04283531],
       [0.00044407, 0.01967537],
       [0.00029226, 0.00788003]])

array([[0., 1.],
       [0., 1.],
       [1., 1.],
       [1., 1.]])

**1.9**: Now, it's just a matter of selecting the most likely option for the last day. For sequence 1, it's hot (1):

In [1532]:
x5 = np.argmax(m[-1])
display(x5)

1

**1.10** With the most likely option, let's no go back and generate the sequence. In this case, we ended up with a sequence of cold, cold, hot, cold, hot for observations 2, 3, 3, 2, 3. 

In [1533]:
sequence = list()
sequence.insert(0, x5)
last_index = x5
for i in range(c.shape[0] - 1, -1, -1):
  curr_index = int(c[i][last_index])
  sequence.insert(0, curr_index)
  last_index = curr_index
display(sequence)

[1, 1, 1, 1, 1]

**1.11**: One last thing! Let's turn this sequence generation code into a function for reuse. 




In [0]:
def generate_sequence(c, x):
  sequence = list()
  sequence.insert(0, x)
  last_index = x
  for i in range(c.shape[0] - 1, -1, -1):
    curr_index = int(c[i][last_index])
    sequence.insert(0, curr_index)
    last_index = curr_index
  return sequence

In [1535]:
generate_sequence(c, x5)

[1, 1, 1, 1, 1]

**1.12**: Now, for fun, let's run the Viterbi algorithm for all ten sequences:

In [0]:
def viterbi_experiment(test, observation, transition):
  final_sequences = []
  for i in range(1, test_dataframe.shape[0] + 1):
    m, c = init_matrices(test_dataframe, observation_dataframe, transition_dataframe, i)
    for j in range(1, m.shape[0]):
      step(m, c, test_dataframe, observation_dataframe, transition_dataframe, i, j)
    x = np.argmax(m[-1])
    final_sequences.append(generate_sequence(c, x))
  return final_sequences

In [0]:
def dump_sequences(sequences):
  for i, s in enumerate(sequences):
    print(f'Sequence {i + 1}: {s}')

In [1538]:
viterbi_sequences = viterbi_experiment(test_dataframe, observation_dataframe, transition_dataframe)
dump_sequences(viterbi_sequences)

Sequence 1: [1, 1, 1, 1, 1]
Sequence 2: [1, 1, 1, 1]
Sequence 3: [0, 0, 0, 0, 0]
Sequence 4: [0, 0, 0]
Sequence 5: [0, 0, 0, 0, 0]
Sequence 6: [0, 0, 0, 0]
Sequence 7: [1, 1, 1]
Sequence 8: [1, 1, 0, 0]
Sequence 9: [0, 1, 1, 1, 1]
Sequence 10: [0, 0, 0]


## Part 2: Likelihood Sampling



Using the same network, implement likelihood sampling for approximate inference.  For any test sequence, sample complete sequences of the hidden states n times, where n can range from 10 to 100000 samples. The goal is to approximate the likelihood of all possible sequences.

Assuming the Viterbi sequence is "correct", how long (how many samples) does it take the sampler to converge so that you get the highest match between samples and the Viterbi sequence?

How do I sample a sequence?  In essence, pick a length (3, 4, or 5) - pick the same lengths as each test sample.  Then, sample each weather-day (Hot/Cold) according to the distribution given by the transition network.  You will need to sample Day 1 before sampling Day 2, for example.  You will then have a complete sample of sequence length 3/4/5).  The weight of that sequence sample will be the product of the observation probabilities given the sample (why?).  You can then judge by the overall weight which the most likely weather sequence would be.  Does the best string match your Viterbi answer?

Note: Technically, in the original problem there is the probability of sampling STOP given either HOT or COLD.  For this section of the homework, please just remove the STOP probability and renormalize the other two probabilities so that they sum to one.

**2.0**: In order to begin sampling, we'll need to write a sampling function.

In [0]:
import random

In [0]:
def update_sequence(cold_probability: float, sample_sequence: list):
  """
  A helper function which appends 0 to the list of cold is more likely than hot
  """
  random_num = random.random()  
  if random_num < cold_probability:
    sample_sequence.append(0)
  else:
    sample_sequence.append(1)

In [0]:
def sample(test: pd.DataFrame, transition: pd.DataFrame, sequence: int):
  """
  A helper function which generates a sample sequence by generating random
  numbers for each day and comparing those numbers with the transition 
  probability.
  """
  sample_sequence = list()
  COLS = test.loc[:, test.columns != SEQUENCE_NUM_LABEL]  
  for index, obs in enumerate(COLS):
    OBS = test_observation(test, COLS, sequence, index)
    if OBS != 0: # Sequence continues
      if index == 0: # Use prior probabilities for first iteration
        PRIOR_COLD = prior_probability(transition, COLD_LABEL)
        update_sequence(PRIOR_COLD, sample_sequence)
      else: # Otherwise, use transition probabilities
        if sample_sequence[index - 1] == 0: # Yesterday was cold
          P_C_TODAY_GIVEN_COLD_YESTERDAY = transition_probability(transition, COLD_LABEL, COLD_LABEL)
          update_sequence(P_C_TODAY_GIVEN_COLD_YESTERDAY, sample_sequence)
        else:  # Yesterday was hot
          P_C_TODAY_GIVEN_HOT_YESTERDAY = transition_probability(transition, COLD_LABEL, HOT_LABEL)
          update_sequence(P_C_TODAY_GIVEN_HOT_YESTERDAY, sample_sequence)
  return sample_sequence

**2.1**: Now that we can generate samples, we'll need to be able to compute a weight for each sample. According to the directions, we should be able to do that by computing the product of the observations probabilities for each element in the sequence:

In [0]:
def weight(test: pd.DataFrame, observations: pd.DataFrame, sample_sequence: list, sequence: int):
  """
  A helper function which computes the weight of a sequence according to the
  obervation probabilities.
  """
  product = 1
  COLS = test.loc[:, test.columns != SEQUENCE_NUM_LABEL]  
  for index, state in enumerate(sample_sequence):
    obs = test_observation(test, COLS, sequence, int(index))
    label = COLD_LABEL if state == 0 else HOT_LABEL
    probability = emission_probability(observations, label, obs)
    product *= probability
  return product

**2.2**: With these functions, we're able to generate a sample sequence using random numbers and the transition probabilities. As a result, we can run experiments for each of the sequences to see if we can converge on the viterbi algorithm. As a result, we should write an experiment function which will do that for a given sequence.

In [0]:
def generate_sample_mapping(test, transition, sequence, n):
  """
  A helper function which counts every sample that's generated.
  """
  sequence_dict = {}
  for i in range(n):
    s = tuple(sample(test, transition, sequence))
    if s in sequence_dict:
      sequence_dict[s] += 1
    else:
      sequence_dict[s] = 1
  return sequence_dict

In [0]:
def compute_best_sample_and_weight(test, observation, sequence, sequence_dict):
  """
  A helper method which computes the best sample and its score.
  """
  best_sample = []
  best_score = 0
  for s, count in sequence_dict.items():
    w = weight(test, observation, s, sequence)
    sequence_dict[s] = sequence_dict[s] * w
    if sequence_dict[s] > best_score:
      best_score = sequence_dict[s]
      best_sample = s
  return best_sample, best_score

In [0]:
def likelihood_experiment(test: pd.DataFrame, observation: pd.DataFrame, transition: pd.DataFrame, n: int):
  """
  A helper method which performs the likelihood experiment N times for all
  observed sequences.
  """
  final_sequences = []
  for i in range(1, test_dataframe.shape[0] + 1):  
    sequence_dict = generate_sample_mapping(test, transition, i, n)
    best_sample, _ = compute_best_sample_and_weight(test, observation, i, sequence_dict)
    final_sequences.append(best_sample)
  return final_sequences

In [0]:
def performance(viterbi_sequences, likelihood_sequences):
  """
  A helper method which calculates performance of the likelihood algorithm
  based on how well it matches the viterbi algorithm for the same data.
  """
  count = 0
  for s1, s2 in zip(viterbi_sequences, likelihood_sequences):
    if tuple(s1) == s2:
      count += 1
  print(f'Performance: {count / len(viterbi_sequences) * 100}% match')

**2.3**: Now, we can run a few experiments to see if we get any matches. But first, we should renormalize the transition probabilities.

In [1547]:
transition_dataframe_copy = transition_dataframe.copy()
base_cold = transition_dataframe_copy[COLD_LABEL][:2].sum()
base_hot = transition_dataframe_copy[HOT_LABEL][:2].sum()
display(base_cold, base_hot)
transition_dataframe_copy[COLD_LABEL] /= base_cold 
transition_dataframe_copy[HOT_LABEL] /= base_hot
transition_dataframe_copy

0.9299999999999999

0.9299999999999999

P(x|...)         C         H  START
0        C  0.924731  0.075269    0.5
1        H  0.075269  0.924731    0.5
2     STOP  0.075269  0.075269    0.0

**2.4**: With the updated transition table, we should be able to run our experiment for each sequence.

In [1548]:
likelihood_sequences = likelihood_experiment(test_dataframe, observation_dataframe, transition_dataframe_copy, 10)
dump_sequences(likelihood_sequences)
performance(viterbi_sequences, likelihood_sequences)

Sequence 1: (1, 1, 1, 1, 1)
Sequence 2: (1, 1, 1, 1)
Sequence 3: (0, 0, 0, 0, 0)
Sequence 4: (0, 0, 0)
Sequence 5: (0, 0, 0, 0, 0)
Sequence 6: (0, 0, 0, 0)
Sequence 7: (1, 1, 1)
Sequence 8: (0, 0, 0, 0)
Sequence 9: (0, 1, 1, 1, 0)
Sequence 10: (0, 0, 0)
Performance: 80.0% match


In [1549]:
likelihood_sequences = likelihood_experiment(test_dataframe, observation_dataframe, transition_dataframe_copy, 100)
dump_sequences(likelihood_sequences)
performance(viterbi_sequences, likelihood_sequences)

Sequence 1: (1, 1, 1, 1, 1)
Sequence 2: (1, 1, 1, 1)
Sequence 3: (0, 0, 0, 0, 0)
Sequence 4: (0, 0, 0)
Sequence 5: (0, 0, 0, 0, 0)
Sequence 6: (0, 0, 0, 0)
Sequence 7: (1, 1, 1)
Sequence 8: (1, 1, 0, 0)
Sequence 9: (0, 1, 1, 1, 1)
Sequence 10: (0, 0, 0)
Performance: 100.0% match


In [1550]:
likelihood_sequences = likelihood_experiment(test_dataframe, observation_dataframe, transition_dataframe_copy, 1000)
dump_sequences(likelihood_sequences)
performance(viterbi_sequences, likelihood_sequences)

Sequence 1: (1, 1, 1, 1, 1)
Sequence 2: (1, 1, 1, 1)
Sequence 3: (0, 0, 0, 0, 0)
Sequence 4: (0, 0, 0)
Sequence 5: (0, 0, 0, 0, 0)
Sequence 6: (0, 0, 0, 0)
Sequence 7: (1, 1, 1)
Sequence 8: (1, 1, 0, 0)
Sequence 9: (0, 1, 1, 1, 1)
Sequence 10: (0, 0, 0)
Performance: 100.0% match


In [1551]:
likelihood_sequences = likelihood_experiment(test_dataframe, observation_dataframe, transition_dataframe_copy, 10000)
dump_sequences(likelihood_sequences)
performance(viterbi_sequences, likelihood_sequences)

Sequence 1: (1, 1, 1, 1, 1)
Sequence 2: (1, 1, 1, 1)
Sequence 3: (0, 0, 0, 0, 0)
Sequence 4: (0, 0, 0)
Sequence 5: (0, 0, 0, 0, 0)
Sequence 6: (0, 0, 0, 0)
Sequence 7: (1, 1, 1)
Sequence 8: (1, 1, 0, 0)
Sequence 9: (0, 1, 1, 1, 1)
Sequence 10: (0, 0, 0)
Performance: 100.0% match


**2.5**: Closing comments! In general, the likelihood estimation technique is very slow, and I suspect my implementation isn't doing me any favors. As a result, I am not able to simulate each sequence for 100,000 samples. Unfortunately, 10,000 takes quite a bit of time. 

That said, I was able to get the likelihood estimation algorithm to converge rather quickly. In fact, 10 samples was all I needed to converge at about 80%. By 1000 samples, I pretty consistently reach convergence. However, I suspect the transition probabilities are a bit generous to me in this case. After all, we are more likely to stay hot two days in a row (i.e. 86%) than switch, so the likelihood estimation picks up on that very quickly.

In general, the most interesting sequences are 8 and 9 because we don't get sweeping temperatures. Naturally, these are the sequences that fail for low samples during the likelihood algorithm because they aren't being generated. As a result, the weights have no chance of playing a role. 

I would be interested in seeing this algorithm play out with different probabilities, but that may be an exercise for another time.